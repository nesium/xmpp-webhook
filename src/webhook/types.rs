use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct Committer {
    pub name: String,
    pub email: Option<String>,
    pub date: Option<String>,
    pub username: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Commit {
    pub id: String,
    pub tree_id: String,
    pub distinct: bool,
    pub message: String,
    pub timestamp: String,
    pub url: String,
    pub author: Committer,
    pub committer: Committer,
    pub added: Vec<String>,
    pub modified: Vec<String>,
    pub removed: Vec<String>,
}

#[derive(Debug, Deserialize)]
pub struct User {
    pub login: String,
    pub id: usize,
    pub name: Option<String>,
    pub email: Option<String>,
    pub html_url: String,
    #[serde(rename = "type")]
    pub type_: UserType,
}

#[derive(Debug, Deserialize)]
pub enum UserType {
    Bot,
    Organization,
    User,
}

#[derive(Debug, Deserialize)]
pub struct Label {
    pub id: usize,
    pub url: String,
    pub name: String,
    pub description: Option<String>,
    pub color: String,
    pub default: bool,
}

#[derive(Debug, Deserialize)]
pub struct Organization {
    pub login: String,
    pub id: usize,
    pub html_url: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct Repository {
    pub id: usize,
    pub name: String,
    pub full_name: String,
    pub private: bool,
    pub owner: User,
    pub html_url: String,
    pub description: Option<String>,
    pub fork: bool,
    pub url: String,
}

pub type MilestoneState = PullRequestState;
#[derive(Debug, Deserialize)]
pub struct Milestone {
    pub url: String,
    pub html_url: String,
    pub id: usize,
    pub number: usize,
    pub title: String,
    pub description: Option<String>,
    pub creator: User,
    pub open_issues: usize,
    pub closed_issues: usize,
    pub state: MilestoneState,
    pub created_at: String,
    pub updated_at: String,
    pub due_on: Option<String>,
    pub closed_at: Option<String>,
}

pub type IssueState = PullRequestState;

#[derive(Debug, Deserialize)]
pub struct Issue {
    pub url: String,
    pub repository_url: String,
    pub html_url: String,
    pub id: usize,
    pub number: usize,
    pub title: String,
    pub user: User,
    pub labels: Option<Vec<Label>>,
    pub state: Option<IssueState>,
    pub locked: Option<bool>,
    pub assignee: Option<User>,
    pub assignees: Vec<User>,
    pub comments: usize,
    pub created_at: String,
    pub updated_at: String,
    pub closed_at: Option<String>,
    pub draft: Option<bool>,
    pub body: Option<String>,
    pub state_reason: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct IssueComment {
    pub url: String,
    pub html_url: String,
    pub issue_url: String,
    pub id: usize,
    pub user: User,
    pub created_at: String,
    pub updated_at: String,
    pub body: String,
}

#[derive(Debug, Deserialize)]
pub struct PushEvent {
    #[serde(rename = "ref")]
    pub ref_: String,
    pub before: String,
    pub after: String,
    pub created: bool,
    pub deleted: bool,
    pub forced: bool,
    pub base_ref: Option<String>,
    pub commits: Vec<Commit>,
    pub head_commit: Option<Commit>,
    pub repository: Repository,
    pub pusher: Committer,
    pub sender: User,
    pub organization: Option<Organization>,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "action")]
#[serde(rename_all = "snake_case")]
pub enum IssueCommentEvent {
    Created(IssueCommentCreatedEvent),
    Deleted(IssueCommentDeletedEvent),
    Edited(IssueCommentEditedEvent),
}

#[derive(Debug, Deserialize)]
#[serde(tag = "action")]
#[serde(rename_all = "snake_case")]
pub enum IssuesEvent {
    Assigned(IssuesAssignedEvent),
    Closed(IssuesClosedEvent),
    Deleted(IssuesDeletedEvent),
    Demilestoned(IssuesDemilestonedEvent),
    Edited(IssuesEditedEvent),
    Labeled(IssuesLabeledEvent),
    Locked(IssuesLockedEvent),
    Milestoned(IssuesMilestonedEvent),
    Opened(IssuesOpenedEvent),
    Pinned(IssuesPinnedEvent),
    Reopened(IssuesReopenedEvent),
    Transferred(IssuesTransferredEvent),
    Unassigned(IssuesUnassignedEvent),
    Unlabeled(IssuesUnlabeledEvent),
    Unlocked(IssuesUnlockedEvent),
    Unpinned(IssuesUnpinnedEvent),
}

#[derive(Debug, Deserialize)]
pub enum PullRequestState {
    #[serde(rename = "closed")]
    Closed,
    #[serde(rename = "open")]
    Open,
}

pub type IssueCommentCreatedEventIssueState = PullRequestState;
#[derive(Debug, Deserialize)]
pub struct IssueCommentCreatedEventIssue {
    pub assignee: Option<User>,
    pub state: IssueCommentCreatedEventIssueState,
    pub locked: bool,
    pub labels: Vec<Label>,
    #[serde(flatten)]
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssueCommentCreatedEvent {
    pub issue: IssueCommentCreatedEventIssue,
    pub comment: IssueComment,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}

pub type IssueCommentDeletedEventIssueState = PullRequestState;
#[derive(Debug, Deserialize)]
pub struct IssueCommentDeletedEventIssue {
    pub assignee: Option<User>,
    pub state: IssueCommentDeletedEventIssueState,
    pub locked: bool,
    pub labels: Vec<Label>,
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssueCommentDeletedEvent {
    pub issue: IssueCommentDeletedEventIssue,
    pub comment: IssueComment,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}

pub type IssueCommentEditedEventIssueState = PullRequestState;
#[derive(Debug, Deserialize)]
pub struct IssueCommentEditedEventIssue {
    pub assignee: Option<User>,
    pub state: IssueCommentEditedEventIssueState,
    pub locked: bool,
    pub labels: Vec<Label>,
    #[serde(flatten)]
    pub issue: Issue,
}

#[derive(Debug, Deserialize)]
pub struct IssueCommentEditedEvent {
    pub issue: IssueCommentEditedEventIssue,
    pub comment: IssueComment,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}

#[derive(Debug, Deserialize)]
pub struct IssuesAssignedEvent {
    pub issue: Issue,
    pub assignee: Option<User>,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesClosedEventIssue {
    pub closed_at: String,
    #[serde(flatten)]
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssuesClosedEvent {
    pub issue: IssuesClosedEventIssue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesDeletedEvent {
    pub issue: Issue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesDemilestonedEventIssue {
    #[serde(flatten)]
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssuesDemilestonedEvent {
    pub issue: IssuesDemilestonedEventIssue,
    pub milestone: Milestone,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesEditedEventChangesBody {
    pub from: String,
}
#[derive(Debug, Deserialize)]
pub struct IssuesEditedEventChangesTitle {
    pub from: String,
}
#[derive(Debug, Deserialize)]
pub struct IssuesEditedEventChanges {
    pub body: Option<IssuesEditedEventChangesBody>,
    pub title: Option<IssuesEditedEventChangesTitle>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesEditedEvent {
    pub issue: Issue,
    pub label: Option<Label>,
    pub changes: IssuesEditedEventChanges,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesLabeledEvent {
    pub issue: Issue,
    pub label: Option<Label>,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}

#[derive(Debug, Deserialize)]
pub enum PullRequestActiveLockReason {
    #[serde(rename = "off-topic")]
    OffTopic,
    #[serde(rename = "resolved")]
    Resolved,
    #[serde(rename = "spam")]
    Spam,
    #[serde(rename = "too heated")]
    TooHeated,
}

pub type IssuesLockedEventIssueActiveLockReason = PullRequestActiveLockReason;
#[derive(Debug, Deserialize)]
pub struct IssuesLockedEventIssue {
    pub active_lock_reason: Option<IssuesLockedEventIssueActiveLockReason>,
    #[serde(flatten)]
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssuesLockedEvent {
    pub issue: IssuesLockedEventIssue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesMilestonedEventIssue {
    pub milestone: Milestone,
    #[serde(flatten)]
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssuesMilestonedEvent {
    pub issue: IssuesMilestonedEventIssue,
    pub milestone: Milestone,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesOpenedEventChanges {
    pub old_issue: Issue,
    pub old_repository: Repository,
}
#[derive(Debug, Deserialize)]
pub struct IssuesOpenedEventIssue {
    pub closed_at: (),
    #[serde(flatten)]
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssuesOpenedEvent {
    pub changes: Option<IssuesOpenedEventChanges>,
    pub issue: IssuesOpenedEventIssue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesPinnedEvent {
    pub issue: Issue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
pub type IssuesReopenedEventIssue = Issue;
#[derive(Debug, Deserialize)]
pub struct IssuesReopenedEvent {
    pub issue: IssuesReopenedEventIssue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesTransferredEventChanges {
    pub new_issue: Issue,
    pub new_repository: Repository,
}
#[derive(Debug, Deserialize)]
pub struct IssuesTransferredEvent {
    pub changes: IssuesTransferredEventChanges,
    pub issue: Issue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesUnassignedEvent {
    pub issue: Issue,
    pub assignee: Option<User>,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesUnlabeledEvent {
    pub issue: Issue,
    pub label: Option<Label>,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesUnlockedEventIssue {
    #[serde(flatten)]
    pub issue: Issue,
}
#[derive(Debug, Deserialize)]
pub struct IssuesUnlockedEvent {
    pub issue: IssuesUnlockedEventIssue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
#[derive(Debug, Deserialize)]
pub struct IssuesUnpinnedEvent {
    pub issue: Issue,
    pub repository: Repository,
    pub sender: User,
    pub organization: Option<Organization>,
}
