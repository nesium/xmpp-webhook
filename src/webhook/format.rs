use github_webhook::payload_types::{IssuesOpenedEvent, PushEvent};

pub fn format_push_event(event: &PushEvent) -> String {
    let commits_markdown: Vec<String> = event
        .commits
        .iter()
        .map(|commit| {
            format!(
                "- **Commit**: [{}]({})\n  **Author**: {} <{}>\n  **Message**: {}\n",
                &commit.id[..7],
                commit.url,
                commit.author.name,
                commit.author.email.unwrap_or("<no email>"),
                commit.message
            )
        })
        .collect();

    format!(
        "New commits pushed to [{}]({})\n\n{}",
        event.repository.name,
        event.repository.html_url,
        commits_markdown.join("\n")
    )
}

pub fn format_issue_opened(event: &IssuesOpenedEvent) -> String {
    format!(
        r#"[{user}]({user_url}) has opened [issue #{issue_number}]({issue_url}) in [{repo}]({repo_url})

**Title**: {issue_title}
        "#,
        user = event.issue.issue.user.login,
        user_url = event.issue.issue.user.html_url,
        repo = event.repository.name,
        repo_url = event.repository.html_url,
        issue_number = event.issue.issue.number,
        issue_title = event.issue.issue.title,
        issue_url = event.issue.issue.html_url
    )
}
