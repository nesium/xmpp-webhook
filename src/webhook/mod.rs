pub use repo_mapping::RepoMapping;
pub use workflow_runs_store::WorkflowRunsStore;

mod repo_mapping;
mod workflow_runs_store;
