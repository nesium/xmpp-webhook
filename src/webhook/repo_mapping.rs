use std::collections::HashMap;
use std::sync::Arc;

use prose_xmpp::BareJid;

use crate::config::RepoSettings;

#[derive(Debug, Clone)]
pub struct RepoMapping(Arc<HashMap<String, BareJid>>);

impl RepoMapping {
    pub fn new(mapping: Vec<RepoSettings>) -> Self {
        Self(Arc::new(
            mapping.into_iter().map(|m| (m.repo, m.room)).collect(),
        ))
    }

    pub fn get(&self, room: &str) -> Option<&BareJid> {
        self.0.get(room)
    }
}
