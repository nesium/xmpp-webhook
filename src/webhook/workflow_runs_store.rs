use std::collections::HashMap;
use std::sync::{Arc, Mutex};

/// A store for managing workflow runs associated with repositories.
#[derive(Debug, Clone)]
pub struct WorkflowRunsStore(Arc<Mutex<HashMap<String, Vec<WorkflowRun>>>>);

impl WorkflowRunsStore {
    pub fn new() -> Self {
        Self(Default::default())
    }

    /// Records a failed workflow run for the specified repository. If the run is already recorded,
    /// it does not add a duplicate.
    ///
    /// # Arguments
    /// * `repo` - The repository name.
    /// * `workflow_id` - A unique identifier for the workflow run.
    /// * `head_branch` - The branch name at the head during the workflow run.
    pub fn workflow_failed(
        &self,
        repo: impl Into<String>,
        workflow_id: u64,
        head_branch: impl Into<String>,
    ) {
        let run = WorkflowRun {
            workflow_id,
            head_branch: head_branch.into(),
        };
        let mut map = self.0.lock().unwrap();
        let runs = map.entry(repo.into()).or_default();

        if runs.contains(&run) {
            return;
        }

        runs.push(run);
    }

    /// Removes workflow runs from the store that match the specified `workflow_id` and `head_branch`.
    /// Returns `true` if any runs were removed, otherwise `false`.
    ///
    /// # Arguments
    /// * `repo` - The repository name.
    /// * `workflow_id` - The unique identifier for the workflow to be removed.
    /// * `head_branch` - The branch name at the head during the workflow run.
    pub fn workflow_succeeded(
        &self,
        repo: impl AsRef<str>,
        workflow_id: u64,
        head_branch: impl AsRef<str>,
    ) -> bool {
        let mut map = self.0.lock().unwrap();
        let Some(runs) = map.get_mut(repo.as_ref()) else {
            return false;
        };

        let runs_len = runs.len();
        let head_branch = head_branch.as_ref();
        runs.retain(|run| run.workflow_id != workflow_id || &run.head_branch != head_branch);

        runs.len() != runs_len
    }
}

#[derive(Debug, PartialEq)]
struct WorkflowRun {
    workflow_id: u64,
    head_branch: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_store_is_empty() {
        let store = WorkflowRunsStore::new();
        let map = store.0.lock().unwrap();
        assert!(map.is_empty());
    }

    #[test]
    fn test_workflow_failed_adds_run() {
        let store = WorkflowRunsStore::new();
        store.workflow_failed("repo1", 1, "main");
        let map = store.0.lock().unwrap();
        assert_eq!(map.get("repo1").unwrap().len(), 1);
    }

    #[test]
    fn test_workflow_failed_does_not_add_duplicate() {
        let store = WorkflowRunsStore::new();
        store.workflow_failed("repo1", 1, "main");
        store.workflow_failed("repo1", 1, "main"); // Attempt to add duplicate
        let map = store.0.lock().unwrap();
        assert_eq!(map.get("repo1").unwrap().len(), 1);
    }

    #[test]
    fn test_workflow_succeeded_removes_correct_run() {
        let store = WorkflowRunsStore::new();
        store.workflow_failed("repo1", 1, "main");
        store.workflow_failed("repo1", 2, "dev");
        let removed = store.workflow_succeeded("repo1", 1, "main");
        let map = store.0.lock().unwrap();
        assert!(removed);
        assert_eq!(map.get("repo1").unwrap().len(), 1);
        assert_eq!(map.get("repo1").unwrap()[0].workflow_id, 2);
    }

    #[test]
    fn test_workflow_succeeded_returns_false_if_no_match() {
        let store = WorkflowRunsStore::new();
        store.workflow_failed("repo1", 1, "main");
        let removed = store.workflow_succeeded("repo1", 999, "main");
        assert!(!removed);
    }
}
