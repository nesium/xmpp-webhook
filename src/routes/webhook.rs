use std::str::FromStr;

use crate::webhook::format;
use actix_web::http::StatusCode;
use actix_web::{web, HttpRequest, HttpResponse, Responder, ResponseError};
use github_webhook::payload_types::{IssueCommentEvent, IssuesEvent, PushEvent};
use prose_xmpp::BareJid;

use crate::xmpp::XMPPHandle;

#[derive(thiserror::Error, Debug)]
pub enum WebhookError {
    #[error("Missing event type")]
    MissingEventType,
    #[error(transparent)]
    DeserializationError(#[from] serde_json::Error),
}

impl ResponseError for WebhookError {
    fn status_code(&self) -> StatusCode {
        match self {
            WebhookError::MissingEventType => StatusCode::BAD_REQUEST,
            WebhookError::DeserializationError(_) => StatusCode::BAD_REQUEST,
        }
    }
}

pub async fn webhook(
    req: HttpRequest,
    xmpp: web::Data<XMPPHandle>,
    body: web::Bytes,
) -> Result<impl Responder, WebhookError> {
    let Some(event_type) = req
        .headers()
        .get("X-GitHub-Event")
        .and_then(|val| val.to_str().ok())
    else {
        return Err(WebhookError::MissingEventType);
    };

    let message = match event_type {
        "push" => {
            let event = serde_json::from_slice::<PushEvent>(&body)?;
            format::format_push_event(&event)
        }
        "issues" => {
            let event = serde_json::from_slice::<IssuesEvent>(&body)?;
            match event {
                IssuesEvent::Opened(event) => format::format_issue_opened(&event),
                IssuesEvent::Closed(event) => format::format_issue_closed(&event),
                IssuesEvent::Reopened(event) => format::format_issue_reopened(&event),
                _ => return Ok(HttpResponse::Ok().body("ok")),
            }
        }
        "issue_comment" => {
            let event = serde_json::from_slice::<IssueCommentEvent>(&body)?;
            match event {
                IssueCommentEvent::Created(event) => format::format_issue_comment_created(&event),
                _ => return Ok(HttpResponse::Ok().body("ok")),
            }
        }
        _ => return Ok(HttpResponse::Ok().body("ok")),
    };

    xmpp.send_message(BareJid::from_str("marc@prose.org").unwrap(), message);
    Ok(HttpResponse::Ok().body("message sent"))
}
