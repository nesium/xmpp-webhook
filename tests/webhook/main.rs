use insta::assert_snapshot;

use xmpp_webhook::webhook::format::{
    format_issue_comment_created, format_issue_opened, format_push_event,
};
use xmpp_webhook::webhook::types::{IssueCommentCreatedEvent, IssuesOpenedEvent, PushEvent};

#[test]
fn test_push() {
    let contents = include_str!("fixtures/push.json");
    assert_snapshot!(format_push_event(
        &serde_json::from_str::<PushEvent>(contents).unwrap()
    ));
}

#[test]
fn test_issue_opened() {
    let contents = include_str!("fixtures/issue_opened.json");
    assert_snapshot!(format_issue_opened(
        &serde_json::from_str::<IssuesOpenedEvent>(contents).unwrap()
    ));
}

#[test]
fn test_issue_comment_created() {
    let contents = include_str!("fixtures/issue_comment_created.json");
    assert_snapshot!(format_issue_comment_created(
        &serde_json::from_str::<IssueCommentCreatedEvent>(contents).unwrap()
    ));
}
