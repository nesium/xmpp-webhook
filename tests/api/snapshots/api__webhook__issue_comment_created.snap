---
source: tests/webhook/main.rs
expression: "format_issue_comment_created(&serde_json::from_str::<IssueCommentCreatedEvent>(contents).unwrap())"
---
[Codertocat](https://github.com/Codertocat) commented on [issue #1](https://github.com/Codertocat/Hello-World/issues/1) in [Hello-World](https://github.com/Codertocat/Hello-World)

**Title**: Spelling error in the README file
